docs: add page for long tasks causing frame drops (#7625)
getsentry/sentry-docs: src/docs/product/issues/issue-details/performance-issues/long-task-causing-frame-drop/index.mdx

---
title: "Long Task Causing Frame Drop"
sidebar_order: 30
redirect_from:
  - /product/issues/performance-issues/long-task-causing-frame-drop/
description: "Learn more about how we detect Long Task Causing Frame Drop issues and what you can do to help fix them."
---

The main (or UI) thread in a mobile app is responsible for handling all user interaction and needs to be able to respond to gestures and taps in real time. If a long-running operation blocks the main thread, the app becomes unresponsive, impacting the quality of the user experience.

We can detect some specific causes for a main thread stall, like decoding [images](LINK) or [JSON](LINK), or [searching strings with regexes](LINK), but there are arbitrarily many things that could cause one. If a main thread stall occurs that doesn't match one of our specific detectors, but is causing your app to drop UI frames, it gets reported under this generic issue type.

## Detection Criteria

> [Profiling](/product/profiling/) must be enabled for Sentry to detect Long Task Causing Frame Drop issues. Sentry looks for profiles that recorded a [frozen](LINK) UI frame, and then searches for the bottom-most function call from application code that persisted for the longest time while the display link waited to draw the frozen frame.

Author: We could also use "deepest" or "leaf function call"

> [Profiling](/product/profiling/) must be enabled for Sentry to detect Long Task Causing Frame Drop issues. Sentry looks for profiles that recorded a [frozen](LINK) UI frame, and then searches for the bottom-most function call from application code that persisted for the longest time while the display link waited to draw the frozen frame.

Author: I reworded this, including some of your suggestions, and used "deepest"

## Function Evidence

> To find additional information about your Long Task Causing Frame Drop problem, go to its **Issue Details** page and scroll down to the "Function Evidence" section, which shows the following:

Reviewer: ```suggestion
To find additional information about your Long Task Causing Frame Drop problem, go to the **Issues** page in Sentry, select the issue you want to examine, and scroll down to the "Function Evidence" section in the "Details" tab, which shows the following:
```

- **Transaction Name:** The name of the transaction where the issue was detected.

> - **Suspect function:** The function that triggered the issue detection.

Reviewer: ```suggestion
- **Suspect Function:** The function that triggered the issue detection.
```

> - **Duration:** How long the function took to execute and the number of consecutive samples collected by the profiler that contained the function.

Reviewer: ```suggestion
- **Duration:** The amount of time it took the function to execute and the number of consecutive samples collected by the profiler that contained the function.
```

![Image Decoding on Main Thread Function Evidence](image-decoding-function-evidence.png)

To view the entire profile associated with the issue, click the “View Profile” button.

> The profile will indicate where the suspect function was called from, along with other functions being called _by_ the suspect function:

Reviewer: ```suggestion
The profile will show where the suspect function was called from, as well as the other functions being called _by_ the suspect function:
```

![Long Task Causing Frame Drop Profile](long-task-frame-drop-profile.png)

## Stack Trace

> The “Stack Trace” section shows a full stack trace for where the detected image decoding function was called from:

Reviewer: ```suggestion
The “Stack Trace” section shows a full stack trace view, including where the detected image decoding function was called from:
```

![Long Task Causing Frame Drop Stack Trace](long-task-frame-drop-stack-trace.png)

## Example

### iOS

#### Objective-C Concurrency

The following code executes a long-running `while` loop on the main thread:

```objective-c
// given an array of number strings...

NSMutableOrderedSet<NSString *> *sortedEvenNumbers = [NSMutableOrderedSet<NSString *> orderedSet];
[numbers enumerateObjectsUsingBlock:^(NSString * _Nonnull obj, NSUInteger __unused idx, BOOL * _Nonnull __unused stop) {
    if (obj.integerValue % 2 == 0) {
        [sortedEvenNumbers addObject:obj];
    }
}];
```

> Performance could be improved by moving the long computation off of the main thread by dispatching it to a lower Quality of Service (QoS) queue (or `NSOperationQueue`):

Reviewer: ```suggestion
Performance may be improved by moving the long computation off of the main thread and dispatching it to a lower Quality of Service (QoS) queue (or `NSOperationQueue`):
```

```objective-c {tabTitle:Dispatch to Background Queue}
// given an array of number strings...

NSMutableOrderedSet<NSString *> *sortedEvenNumbers = [NSMutableOrderedSet<NSString *> orderedSet];
dispatch_async(dispatch_get_global_queue(QOS_CLASS_UTILITY, 0), ^{
    [numbers enumerateObjectsUsingBlock:^(NSString * _Nonnull obj, NSUInteger __unused idx, BOOL * _Nonnull __unused stop) {
        if (obj.integerValue % 2 == 0) {
            [sortedEvenNumbers addObject:obj];
        }
    }];
});
```

> Specifically for loops, another avenue to consider is to parallelize iterations. There are several options for doing this:

Reviewer: ```suggestion
Another avenue to consider for loops is to parallelize iterations. There are several options for doing this:
```

1. Use `dispatch_apply` to perform iterations of a general loop on a concurrent queue.

```objective-c {tabTitle:Foundation Collection Concurrent Enumeration}
// given an array of number strings...

NSMutableOrderedSet<NSString *> *sortedEvenNumbers = [NSMutableOrderedSet<NSString *> orderedSet];
[numbers enumerateObjectsWithOptions:NSEnumerationConcurrent usingBlock:^(NSString * _Nonnull obj, NSUInteger __unused idx, BOOL * _Nonnull __unused stop) {
    if (obj.integerValue % 2 == 0) {
        [sortedEvenNumbers addObject:obj];
    }
}];
```

```objective-c {tabTitle:dispatch_apply}
// given an array of number strings...

NSMutableOrderedSet<NSString *> *sortedEvenNumbers = [NSMutableOrderedSet<NSString *> orderedSet];
dispatch_apply(numberOfNumbers, dispatch_get_global_queue(QOS_CLASS_UTILITY, 0), ^(size_t iteration) {
    NSString *number = numbers[iteration];
    if (number.integerValue % 2 == 0) {
        [sortedEvenNumbers addObject:number];
    }
});
```

> Keep in mind that concurrent solutions require analysis of thread safety for the code you're parallelizing. You may need to `@synchronize` critical sections, use semaphores, or dispatch back to a serial queue (or the main queue for UI work). Note that both `enumerateObjects...` and `dispatch_apply` are synchronous, so will not return until all iterations have completed; dispatch their invocation asynchronously off the main queue to avoid waiting.

Reviewer: ```suggestion
Keep in mind that concurrent solutions require analysis of thread safety for the code you're parallelizing. You may need to `@synchronize` critical sections, use semaphores, or dispatch back to a serial queue (or the main queue for UI work). Note that both `enumerateObjects...` and `dispatch_apply` are synchronous and won't return until all iterations have completed. Dispatch their invocation asynchronously off the main queue to avoid waiting.
```
